tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -17,134 +17,267 @@
 #include "compression/array.h"
 #include "compression/compression.h"
+#include "nodes/decompress_chunk/batch_merge.h"
 #include "nodes/decompress_chunk/decompress_chunk.h"
 #include "nodes/decompress_chunk/exec.h"
 #include "nodes/decompress_chunk/planner.h"
 #include "ts_catalog/hypertable_compression.h"
 
-typedef enum DecompressChunkColumnType
-{
-	SEGMENTBY_COLUMN,
-	COMPRESSED_COLUMN,
-	COUNT_COLUMN,
-	SEQUENCE_NUM_COLUMN,
-} DecompressChunkColumnType;
-
-typedef struct DecompressChunkColumnState
-{
-	DecompressChunkColumnType type;
-	Oid typid;
-
-	/*
-	 * Attno of the decompressed column in the output of DecompressChunk node.
-	 * Negative values are special columns that do not have a representation in
-	 * the uncompressed chunk, but are still used for decompression. They should
-	 * have the respective `type` field.
-	 */
-	AttrNumber output_attno;
-
-	/*
-	 * Attno of the compressed column in the input compressed chunk scan.
-	 */
-	AttrNumber compressed_scan_attno;
-
-	union
-	{
-		struct
-		{
-			Datum value;
-			bool isnull;
-			int count;
-		} segmentby;
-		struct
-		{
-			DecompressionIterator *iterator;
-		} compressed;
-	};
-} DecompressChunkColumnState;
-
-typedef struct DecompressChunkState
-{
-	CustomScanState csstate;
-	List *decompression_map;
-	List *is_segmentby_column;
-	int num_columns;
-	DecompressChunkColumnState *columns;
-
-	bool initialized;
-	bool reverse;
-	int hypertable_id;
-	Oid chunk_relid;
-	List *hypertable_compression_info;
-	int total_batch_rows;
-	int current_batch_row;
-	MemoryContext per_batch_context;
-} DecompressChunkState;
+/* The value for an invalid batch id */
+#define INVALID_BATCH_ID -1
 
 static TupleTableSlot *decompress_chunk_exec(CustomScanState *node);
 static void decompress_chunk_begin(CustomScanState *node, EState *estate, int eflags);
 static void decompress_chunk_end(CustomScanState *node);
 static void decompress_chunk_rescan(CustomScanState *node);
-static TupleTableSlot *decompress_chunk_create_tuple(DecompressChunkState *state);
+static void decompress_chunk_explain(CustomScanState *node, List *ancestors, ExplainState *es);
+static void decompress_chunk_create_tuple(DecompressChunkState *chunk_state,
+										  DecompressBatchState *batch_state);
+
+static void decompress_initialize_batch_state(DecompressChunkState *chunk_state,
+											  DecompressBatchState *batch_state);
 
 static CustomExecMethods decompress_chunk_state_methods = {
 	.BeginCustomScan = decompress_chunk_begin,
 	.ExecCustomScan = decompress_chunk_exec,
 	.EndCustomScan = decompress_chunk_end,
 	.ReScanCustomScan = decompress_chunk_rescan,
+	.ExplainCustomScan = decompress_chunk_explain,
 };
 
+/*
+ * Build the sortkeys data structure from the list structure in the
+ * custom_private field of the custom scan. This sort info is used to sort
+ * binary heap used for batch merge append.
+ */
+static void
+build_batch_merge_append_sort_info(DecompressChunkState *chunk_state, List *sortinfo)
```

Comment:
I'd call it "batch sorted merge". Sorted merge is a standard term for this algorithm, and "merge append" is a postgres-specific term for a particular scan plan over a partitioned table. I think this might be a little confusing.

Reply:
Thanks, renamed